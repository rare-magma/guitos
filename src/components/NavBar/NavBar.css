.brand {
  color: var(--textcolor);
}

.btn-go-button {
  background-color: var(--lightbgcolor);
  border: 1px solid var(--comment);
  color: var(--purple);
}

.btn-go-button:focus-visible,
.btn-go-button:hover {
  background-color: var(--purple);
  border: 1px solid var(--purple);
  color: var(--lightbgcolor);
}

.btn-outline-danger {
  background-color: var(--lightbgcolor);
  border: 1px solid var(--comment);
  border-radius: 0.375rem;
  color: var(--red);
}

.btn-outline-danger:focus-visible,
.btn-outline-danger:hover {
  background-color: var(--red);
  border: 1px solid var(--red);
  border-radius: 0.375rem;
  color: var(--lightbgcolor);
}

.btn-outline-info {
  background-color: var(--lightbgcolor);
  border: 1px solid var(--comment);
  border-radius: 0.375rem;
  color: var(--cyan);
}

.btn-outline-info:focus-visible,
.btn-outline-info:hover {
  background-color: var(--cyan);
  border: 1px solid var(--cyan);
  border-radius: 0.375rem;
  color: var(--lightbgcolor);
}

/* Search input placeholder */
input[type="text"]::placeholder {
  color: var(--comment);
  opacity: 1;
}

.navbar,
.offcanvas-body,
.offcanvas-header {
  background: var(--bgcolor);
  color: var(--textcolor);
}

.navbar .navbar-toggler-icon {
  background-image: url("data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 100 100'><text y='.9em' font-size='85'>ðŸ’¸</text></svg>");
}

.navbar-toggler {
  border: 1px solid var(--comment);
  border-radius: 0.375rem;
}

.version {
  color: var(--selection);
  font-family:
    ui-monospace,
    SF Mono,
    Menlo,
    Monaco,
    Andale Mono,
    monospace;
  font-size: 0.9em;
}

.rbt-menu > .dropdown-item {
  text-align: start;
  overflow: visible;
}

.dropdown-item.active,
.dropdown-item:active {
  background: var(--lightbgcolor);
  border: 1px solid var(--pink);
  box-shadow: 0 0 0;
  color: var(--textcolor);
}
