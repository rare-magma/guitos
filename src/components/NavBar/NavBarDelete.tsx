import { Button, Nav, OverlayTrigger, Popover, Tooltip } from "react-bootstrap";
import { BsXLg } from "react-icons/bs";
import { useBudget } from "../../context/BudgetContext";

interface NavBarDeleteProps {
  deleteButtonRef: React.RefObject<HTMLButtonElement>;
  handleRemove: (i: string) => void;
  expanded: boolean;
}

export function NavBarDelete({
  deleteButtonRef,
  handleRemove,
  expanded,
}: NavBarDeleteProps) {
  const { budget } = useBudget();

  return (
    <Nav className="m-2">
      <OverlayTrigger
        trigger="click"
        key="nav-deletion-overlay"
        placement="bottom"
        rootClose={true}
        overlay={
          <Popover id={`nav-popover-delete-button`}>
            <Popover.Body>
              <OverlayTrigger
                delay={250}
                placement="bottom"
                overlay={
                  <Tooltip
                    id={`nav-tooltip-delete-budget`}
                    style={{ position: "fixed" }}
                  >
                    delete budget
                  </Tooltip>
                }
              >
                <Button
                  id={"budget-deletion-button"}
                  aria-label="confirm budget deletion"
                  key={"budget-deletion-button"}
                  variant="delete"
                  type="button"
                  ref={deleteButtonRef}
                  onClick={() => budget?.id && handleRemove(budget.id)}
                >
                  {expanded ? "delete budget" : <BsXLg aria-hidden />}
                </Button>
              </OverlayTrigger>
            </Popover.Body>
          </Popover>
        }
      >
        <Button
          className="w-100"
          aria-label="delete budget"
          variant="outline-danger"
          onClick={() => {
            setTimeout(() => {
              deleteButtonRef.current?.focus();
            }, 0);
          }}
        >
          {expanded ? "delete" : <BsXLg aria-hidden />}
        </Button>
      </OverlayTrigger>
    </Nav>
  );
}
