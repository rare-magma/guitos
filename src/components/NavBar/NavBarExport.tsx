import {
  Button,
  InputGroup,
  Nav,
  OverlayTrigger,
  Popover,
  Tooltip,
} from "react-bootstrap";
import { BsDownload } from "react-icons/bs";

interface NavBarExportProps {
  exportButtonRef: React.RefObject<HTMLButtonElement>;
  exportJSONButtonRef: React.RefObject<HTMLButtonElement>;
  handleExport: (i: string) => void;
  expanded: boolean;
}

export function NavBarExport({
  exportButtonRef,
  exportJSONButtonRef,
  handleExport,
  expanded,
}: NavBarExportProps) {
  return (
    <Nav className="m-2">
      <OverlayTrigger
        trigger="click"
        key="nav-export-overlay"
        placement="bottom"
        rootClose={true}
        overlay={
          <Popover id={`nav-popover-export-button`}>
            <Popover.Body>
              <OverlayTrigger
                delay={250}
                placement="bottom"
                overlay={
                  <Tooltip
                    id={`nav-tooltip-export-budget`}
                    style={{ position: "fixed" }}
                  >
                    export budget
                  </Tooltip>
                }
              >
                <InputGroup
                  size="sm"
                  className="mb-1"
                  key={`export-button-group`}
                >
                  <Button
                    id={"budget-export-csv-button"}
                    aria-label="export budget as csv"
                    key={"budget-export-csv-button"}
                    variant="outline-info"
                    type="button"
                    onClick={() => {
                      handleExport("csv");
                    }}
                  >
                    CSV
                  </Button>
                  <Button
                    id={"budget-export-json-button"}
                    aria-label="export budget as json"
                    key={"budget-export-json-button"}
                    variant="outline-info"
                    type="button"
                    ref={exportJSONButtonRef}
                    onClick={() => {
                      handleExport("json");
                    }}
                  >
                    JSON
                  </Button>
                </InputGroup>
              </OverlayTrigger>
            </Popover.Body>
          </Popover>
        }
      >
        <Button
          className="w-100"
          aria-label="export budget"
          ref={exportButtonRef}
          variant="outline-info"
          onClick={() => {
            setTimeout(() => {
              exportJSONButtonRef.current?.focus();
            }, 0);
          }}
        >
          {expanded ? "export" : <BsDownload aria-hidden />}
        </Button>
      </OverlayTrigger>
    </Nav>
  );
}
