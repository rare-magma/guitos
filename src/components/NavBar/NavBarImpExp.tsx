import { Dispatch, SetStateAction, useRef } from "react";
import {
  Button,
  Form,
  InputGroup,
  Nav,
  OverlayTrigger,
  Popover,
  Stack,
  Tooltip,
} from "react-bootstrap";
import { useHotkeys } from "react-hotkeys-hook";
import { BsArrowDownUp, BsUpload } from "react-icons/bs";
import { useBudget } from "../../context/BudgetContext";
import { useDB } from "../../hooks/useDB";
import { budgetToCsv } from "../../utils";

interface NavBarImpExpProps {
  expanded: boolean;
  setExpanded: Dispatch<SetStateAction<boolean>>;
}

export function NavBarImpExp({ expanded, setExpanded }: NavBarImpExpProps) {
  const { budget, budgetList, budgetNameList } = useBudget();
  const { handleImport } = useDB();
  const importRef =
    useRef<HTMLInputElement>() as React.MutableRefObject<HTMLInputElement>;
  const importButtonRef = useRef<HTMLButtonElement>(null);
  const exportCSVButtonRef = useRef<HTMLButtonElement>(null);
  const impExpButtonRef = useRef<HTMLButtonElement>(null);
  const hasOneOrMoreBudgets = budgetNameList && budgetNameList.length > 0;

  useHotkeys("o", (e) => !e.repeat && impExpButtonRef.current?.click(), {
    preventDefault: true,
  });

  useHotkeys("s", (e) => !e.repeat && handleExportJSON(), {
    preventDefault: true,
  });

  useHotkeys("d", (e) => !e.repeat && handleExportCSV(), {
    preventDefault: true,
  });

  function handleExportJSON() {
    if (budget) {
      const date = new Date().toISOString();
      const filename = `guitos-${date.slice(0, -5)}.json`;
      const url = window.URL.createObjectURL(
        new Blob([JSON.stringify(budgetList)]),
      );
      const link = document.createElement("a");
      link.href = url;
      link.setAttribute("download", filename);
      document.body.appendChild(link);
      link.click();
    }
  }

  function handleExportCSV() {
    if (budget) {
      const filename = `${budget.name}.csv`;
      const url = window.URL.createObjectURL(new Blob([budgetToCsv(budget)]));
      const link = document.createElement("a");
      link.href = url;
      link.setAttribute("download", filename);
      document.body.appendChild(link);
      link.click();
    }
  }

  return hasOneOrMoreBudgets ? (
    <Nav className="m-2">
      <OverlayTrigger
        trigger="click"
        key="nav-imp-exp-overlay"
        placement="bottom"
        rootClose={true}
        overlay={
          <Popover id={`nav-popover-imp-exp-button`}>
            <Popover.Body>
              <Stack gap={3}>
                <Stack className="align-self-center" direction="horizontal">
                  <InputGroup
                    size="sm"
                    className="mb-1"
                    key={`export-button-group`}
                  >
                    <Button
                      aria-label="import budget"
                      variant="outline-primary"
                      type="button"
                      ref={importButtonRef}
                      onClick={() => importRef.current?.click()}
                    >
                      import
                    </Button>
                    <Form.Control
                      data-testid="import-form-control"
                      className="straight-corners"
                      type="file"
                      multiple
                      ref={importRef}
                      onChange={(
                        event: React.ChangeEvent<HTMLInputElement>,
                      ) => {
                        setExpanded(false);
                        handleImport(event);
                      }}
                      style={{ display: "none" }}
                    />
                    <Button
                      id={"budget-export-csv-button"}
                      aria-label="export budget as csv"
                      key={"budget-export-csv-button"}
                      variant="outline-primary"
                      type="button"
                      ref={exportCSVButtonRef}
                      style={{
                        minWidth: "7ch",
                      }}
                      onClick={handleExportCSV}
                    >
                      csv
                    </Button>
                    <Button
                      id={"budget-export-json-button"}
                      aria-label="export budget as json"
                      key={"budget-export-json-button"}
                      variant="outline-primary"
                      type="button"
                      style={{
                        minWidth: "7ch",
                      }}
                      onClick={handleExportJSON}
                    >
                      json
                    </Button>
                  </InputGroup>
                </Stack>
              </Stack>
            </Popover.Body>
          </Popover>
        }
      >
        <Button
          className="w-100"
          aria-label="import or export budget"
          variant="outline-primary"
          ref={impExpButtonRef}
          onClick={() => {
            setTimeout(() => {
              importButtonRef.current?.focus();
            }, 0);
          }}
        >
          {expanded ? "import/export" : <BsArrowDownUp aria-hidden />}
        </Button>
      </OverlayTrigger>
    </Nav>
  ) : (
    <Nav className="m-2">
      <OverlayTrigger
        delay={250}
        placement="bottom"
        overlay={
          <Tooltip id={`tooltip-import-budget`} style={{ position: "fixed" }}>
            import budget
          </Tooltip>
        }
      >
        <Form.Group controlId="import">
          <Button
            className="w-100"
            aria-label="import budget"
            variant="outline-primary"
            onClick={() => importRef.current?.click()}
          >
            {expanded ? "import" : <BsUpload aria-hidden />}
          </Button>
          <Form.Control
            data-testid="import-form-control"
            type="file"
            multiple
            ref={importRef}
            onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
              setExpanded(false);
              handleImport(event);
            }}
            style={{ display: "none" }}
          />
        </Form.Group>
      </OverlayTrigger>
    </Nav>
  );
}
