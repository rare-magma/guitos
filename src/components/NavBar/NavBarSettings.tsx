import { useRef } from "react";
import {
  Button,
  InputGroup,
  Nav,
  OverlayTrigger,
  Popover,
  Stack,
  Tooltip,
} from "react-bootstrap";
import { Typeahead } from "react-bootstrap-typeahead";
import { Option } from "react-bootstrap-typeahead/types/types";
import { useHotkeys } from "react-hotkeys-hook";
import { BsGear } from "react-icons/bs";
import { useConfig } from "../../context/ConfigContext";
import { currenciesList } from "../../lists/currenciesList";

interface NavBarSettingsProps {
  expanded: boolean;
}

export function NavBarSettings({ expanded }: NavBarSettingsProps) {
  const { currency, handleCurrency } = useConfig();
  const settingsButtonRef = useRef<HTMLButtonElement>(null);
  const versionRef = useRef<HTMLAnchorElement>(null);

  useHotkeys("t", (e) => !e.repeat && settingsButtonRef.current?.click(), {
    preventDefault: true,
  });

  return (
    <Nav className="m-2">
      <OverlayTrigger
        trigger="click"
        key="nav-settings-overlay"
        placement="bottom"
        rootClose={true}
        overlay={
          <Popover id={`nav-popover-settings-button`}>
            <Popover.Body>
              <Stack gap={3}>
                <Stack className="align-self-center" direction="horizontal">
                  <InputGroup
                    size="sm"
                    className="mb-1"
                    key={`export-button-group`}
                  >
                    <Typeahead
                      id="currency-option-list"
                      maxResults={currenciesList.length}
                      maxHeight="30vh"
                      paginate={false}
                      inputProps={{
                        "aria-label": "select display currency",
                      }}
                      onChange={(c: Option[]) => {
                        if (currenciesList.includes(c[0] as string)) {
                          handleCurrency(c[0] as string);
                        }
                      }}
                      className="w-100 fixed-width-font"
                      style={
                        expanded
                          ? {
                              maxWidth: "10ch",
                              minWidth: "10ch",
                            }
                          : {
                              maxWidth: "7ch",
                              minWidth: "7ch",
                            }
                      }
                      options={currenciesList.sort((a, b) =>
                        a.localeCompare(b),
                      )}
                      placeholder={currency}
                    />
                  </InputGroup>
                </Stack>
                <OverlayTrigger
                  delay={250}
                  placement="bottom"
                  overlay={
                    <Tooltip
                      id={`tooltip-guitos-version`}
                      style={{ position: "fixed" }}
                    >
                      guitos version
                    </Tooltip>
                  }
                >
                  <a
                    className="version align-self-center"
                    aria-label="open guitos changelog"
                    href="https://github.com/rare-magma/guitos/blob/main/CHANGELOG.md"
                    target="_blank"
                    ref={versionRef}
                    rel="noreferrer"
                  >
                    v{APP_VERSION}
                  </a>
                </OverlayTrigger>
              </Stack>
            </Popover.Body>
          </Popover>
        }
      >
        <Button
          className="w-100"
          aria-label="budget settings"
          variant="outline-info"
          ref={settingsButtonRef}
          onClick={() => {
            setTimeout(() => {
              versionRef.current?.focus();
            }, 0);
          }}
        >
          {expanded ? "settings" : <BsGear aria-hidden />}
        </Button>
      </OverlayTrigger>
    </Nav>
  );
}
