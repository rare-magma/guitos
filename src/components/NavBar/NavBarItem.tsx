import { ReactNode } from "react";
import { Button, Nav, OverlayTrigger, Tooltip } from "react-bootstrap";
import "./NavBar.css";

interface NavItemProps {
  itemClassName: string;
  onClick: () => void;
  tooltipID: string;
  tooltipText: string;
  buttonAriaLabel: string;
  buttonClassName?: string;
  buttonVariant: string;
  buttonLink?: string;
  buttonIcon: ReactNode;
  target?: string;
}

export function NavBarItem({
  itemClassName,
  onClick,
  tooltipID,
  tooltipText,
  buttonAriaLabel,
  buttonClassName,
  buttonVariant,
  buttonLink,
  buttonIcon,
  target,
}: NavItemProps) {
  return (
    <Nav.Item
      className={itemClassName}
      onClick={() => {
        onClick();
      }}
    >
      <OverlayTrigger
        delay={250}
        placement="bottom"
        overlay={
          <Tooltip id={tooltipID} style={{ position: "fixed" }}>
            {tooltipText}
          </Tooltip>
        }
      >
        <Button
          aria-label={buttonAriaLabel}
          className={buttonClassName}
          variant={buttonVariant}
          href={buttonLink}
          target={target}
        >
          {buttonIcon}
        </Button>
      </OverlayTrigger>
    </Nav.Item>
  );
}
