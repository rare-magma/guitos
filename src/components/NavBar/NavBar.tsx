import { useEffect, useRef, useState } from "react";
import { Offcanvas, OverlayTrigger, Tooltip } from "react-bootstrap";
import { AsyncTypeahead } from "react-bootstrap-typeahead";
import "react-bootstrap-typeahead/css/Typeahead.bs5.css";
import "react-bootstrap-typeahead/css/Typeahead.css";
import TypeaheadRef from "react-bootstrap-typeahead/types/core/Typeahead";
import { Option } from "react-bootstrap-typeahead/types/types";
import Button from "react-bootstrap/Button";
import Container from "react-bootstrap/Container";
import Form from "react-bootstrap/Form";
import Nav from "react-bootstrap/Nav";
import Navbar from "react-bootstrap/Navbar";
import { useHotkeys } from "react-hotkeys-hook";
import {
  BsArrowClockwise,
  BsArrowCounterclockwise,
  BsArrowLeft,
  BsArrowRight,
  BsPlusLg,
  BsQuestionLg,
  BsXLg,
} from "react-icons/bs";
import { FaRegClone } from "react-icons/fa";
import { useBudget } from "../../context/BudgetContext";
import { useDB } from "../../hooks/useDB";
import { useMove } from "../../hooks/useMove";
import { focusRef, getLabelKey } from "../../utils";
import "./NavBar.css";
import { NavBarDelete } from "./NavBarDelete";
import { NavBarImpExp } from "./NavBarImpExp";
import { NavBarItem } from "./NavBarItem";
import { NavBarSettings } from "./NavBarSettings";

export interface SearchOption {
  id: string;
  item: string;
  name: string;
}

export function NavBar() {
  const searchRef = useRef<TypeaheadRef>(null);
  const nameRef =
    useRef<HTMLInputElement>() as React.MutableRefObject<HTMLInputElement>;
  const deleteButtonRef = useRef<HTMLButtonElement>(null);

  const [expanded, setExpanded] = useState(false);
  const [theme, setTheme] = useState("light");
  const [invalidName, setInvalidName] = useState(false);

  const { select, goBack, canGoBack, goForward, canGoForward, goHome } =
    useMove();
  const {
    options,
    createBudget,
    cloneBudget,
    deleteBudget,
    renameBudget,
    searchBudgets,
  } = useDB();

  const { budget, undo, redo, canRedo, canUndo, budgetNameList } = useBudget();

  const shouldShowBrand = budgetNameList && budgetNameList.length < 1;
  const hasOneOrMoreBudgets = budgetNameList && budgetNameList.length > 0;
  const hasMultipleBudgets = budgetNameList && budgetNameList.length > 1;

  useHotkeys("pageup", (e) => !e.repeat && goForward(), {
    preventDefault: true,
  });

  useHotkeys("pagedown", (e) => !e.repeat && goBack(), {
    preventDefault: true,
  });

  useHotkeys("Home", (e) => !e.repeat && goHome(), {
    preventDefault: true,
  });

  useHotkeys(
    ["/", "f"],
    (e) =>
      !e.repeat &&
      focusRef(
        searchRef as unknown as React.MutableRefObject<HTMLInputElement>,
      ),
    { preventDefault: true },
  );

  useHotkeys("n", (e) => !e.repeat && focusRef(nameRef), {
    preventDefault: true,
  });

  useEffect(() => {
    const mediaQueryList = window.matchMedia("(prefers-color-scheme: dark)");

    if (mediaQueryList.matches) {
      setTheme("dark");
    }

    mediaQueryList.addEventListener("change", (event) =>
      setTheme(event.matches ? "dark" : "light"),
    );
  }, []);

  function handleRename(name: React.ChangeEvent<HTMLInputElement>) {
    if (checkIsUniqueName(name)) {
      renameBudget(name);
    }
  }
  function checkIsUniqueName(name: React.ChangeEvent<HTMLInputElement>) {
    const isUnique = !budgetNameList?.some(
      (b: SearchOption) => b.name === name.target.value,
    );
    setInvalidName(!isUnique);
    return isUnique;
  }

  function handleRemoveBudget(initialId?: string | null) {
    if (initialId) {
      setExpanded(false);
      deleteBudget(initialId);
    }
  }

  function handleSelectAction(option: Option[]) {
    setExpanded(false);
    select(option as SearchOption[]);
    if (searchRef.current) {
      searchRef.current.clear();
    }
  }

  return (
    <Navbar
      variant={theme}
      expand="md"
      onToggle={() => setExpanded(!expanded)}
      data-testid="header"
    >
      <Container fluid className="flex-row">
        {shouldShowBrand && (
          <OverlayTrigger
            delay={250}
            placement="bottom"
            overlay={
              <Tooltip id={`tooltip-guitos-repo`} style={{ position: "fixed" }}>
                view source code in new tab
              </Tooltip>
            }
          >
            <Navbar.Brand className="flex-sm-row">
              <a
                className="brand"
                aria-label="open guitos repository"
                href="https://github.com/rare-magma/guitos"
                target="_blank"
                rel="noreferrer"
              >
                guitos
              </a>
            </Navbar.Brand>
          </OverlayTrigger>
        )}

        <Nav className="flex-row flex-grow-1">
          {budget?.name && (
            <Nav className="flex-row">
              {hasMultipleBudgets && (
                <>
                  <NavBarItem
                    itemClassName={"me-1 my-2"}
                    onClick={goBack}
                    disabled={!canGoBack}
                    tooltipID={"tooltip-go-to-older-budget"}
                    tooltipText={"go to older budget"}
                    buttonAriaLabel={"go to older budget"}
                    buttonVariant={"go-button"}
                    buttonIcon={<BsArrowLeft aria-hidden />}
                  />
                  <NavBarItem
                    itemClassName={"m-2"}
                    onClick={goForward}
                    disabled={!canGoForward}
                    tooltipID={"tooltip-go-to-newer-budget"}
                    tooltipText={"go to newer budget"}
                    buttonAriaLabel={"go to newer budget"}
                    buttonVariant={"go-button"}
                    buttonIcon={<BsArrowRight aria-hidden />}
                  />
                </>
              )}
              <Nav.Item className="m-2 me-3">
                <OverlayTrigger
                  delay={250}
                  placement="bottom"
                  overlay={
                    <Tooltip
                      id={`tooltip-budget-name`}
                      style={{ position: "fixed" }}
                    >
                      budget name
                    </Tooltip>
                  }
                >
                  <>
                    <Form.Control
                      id="budget-name"
                      aria-label={"budget name"}
                      key={"budget-name-key-" + budget.id}
                      defaultValue={budget.name}
                      ref={nameRef}
                      onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
                        handleRename(e);
                      }}
                      style={expanded ? {} : { minWidth: "12ch" }}
                      type="text"
                      maxLength={25}
                      isInvalid={invalidName}
                    />
                    <Form.Control.Feedback tooltip type="invalid">
                      This name is already used by another budget.
                    </Form.Control.Feedback>
                  </>
                </OverlayTrigger>
              </Nav.Item>
            </Nav>
          )}
          <Nav className="flex-grow-1">
            <Navbar.Toggle
              className="ms-auto ms-2 my-2"
              aria-controls={`offcanvasNavbar-expand-md`}
            />
          </Nav>
        </Nav>

        <Navbar.Offcanvas
          id={`offcanvasNavbar-expand-md`}
          placement="end"
          show={expanded}
        >
          <Offcanvas.Header>
            <Offcanvas.Title id={`offcanvasNavbarLabel-expand-md`}>
              {budget?.name ?? "guitos"}
            </Offcanvas.Title>
            <Button
              aria-label="close navigation"
              variant="outline-secondary"
              onClick={() => setExpanded(false)}
            >
              {<BsXLg aria-hidden />}
            </Button>
          </Offcanvas.Header>
          <Offcanvas.Body className="justify-content-end">
            <Nav>
              <Nav className="m-2">
                {hasMultipleBudgets && (
                  <AsyncTypeahead
                    id="search-budget-list"
                    inputProps={{
                      "aria-label": "search in budgets",
                    }}
                    filterBy={["name", "item"]}
                    labelKey={getLabelKey}
                    ref={searchRef}
                    style={expanded ? {} : { minWidth: "14ch" }}
                    onChange={(option: Option[]) => handleSelectAction(option)}
                    className="w-100"
                    options={options}
                    placeholder="Search..."
                    isLoading={false}
                    onSearch={searchBudgets}
                  />
                )}
              </Nav>
            </Nav>
            <Nav>
              {hasOneOrMoreBudgets && (
                <>
                  <NavBarItem
                    disabled={!canUndo}
                    itemClassName={"m-2"}
                    onClick={undo}
                    tooltipID={"tooltip-undo-history"}
                    tooltipText={"undo"}
                    buttonAriaLabel={"undo change"}
                    buttonClassName="w-100"
                    buttonVariant={"outline-info"}
                    buttonIcon={
                      expanded ? (
                        "undo"
                      ) : (
                        <BsArrowCounterclockwise aria-hidden />
                      )
                    }
                  />

                  <NavBarItem
                    disabled={!canRedo}
                    itemClassName={"m-2"}
                    onClick={redo}
                    tooltipID={"tooltip-redo-history"}
                    tooltipText={"redo"}
                    buttonAriaLabel={"redo change"}
                    buttonClassName="w-100"
                    buttonVariant={"outline-info"}
                    buttonIcon={
                      expanded ? "redo" : <BsArrowClockwise aria-hidden />
                    }
                  />
                </>
              )}
              <NavBarItem
                itemClassName={"m-2"}
                onClick={() => {
                  setExpanded(false);
                  createBudget();
                }}
                tooltipID={"tooltip-new-budget"}
                tooltipText={"new budget"}
                buttonAriaLabel={"new budget"}
                buttonClassName="w-100"
                buttonVariant={"outline-success"}
                buttonIcon={expanded ? "new" : <BsPlusLg aria-hidden />}
              />
              {hasOneOrMoreBudgets && (
                <>
                  <NavBarItem
                    itemClassName={"m-2"}
                    onClick={() => {
                      setExpanded(false);
                      cloneBudget();
                    }}
                    tooltipID={"tooltip-clone-budget"}
                    tooltipText={"clone budget"}
                    buttonAriaLabel={"clone budget"}
                    buttonClassName="w-100"
                    buttonVariant={"outline-success"}
                    buttonIcon={expanded ? "clone" : <FaRegClone aria-hidden />}
                  />
                  <NavBarDelete
                    deleteButtonRef={deleteButtonRef}
                    handleRemove={() => handleRemoveBudget(budget?.id)}
                    expanded={expanded}
                  />
                </>
              )}

              <NavBarImpExp expanded={expanded} setExpanded={setExpanded} />

              {hasOneOrMoreBudgets && <NavBarSettings expanded={expanded} />}

              <NavBarItem
                itemClassName={"m-2"}
                onClick={() => undefined}
                tooltipID={"tooltip-guitos-instructions"}
                tooltipText={"open instructions in new tab"}
                buttonAriaLabel={"open instructions in new tab"}
                buttonClassName="w-100"
                buttonVariant={"outline-info"}
                buttonLink="https://github.com/rare-magma/guitos#getting-started"
                buttonIcon={
                  expanded ? "instructions" : <BsQuestionLg aria-hidden />
                }
                target="_blank"
              />
            </Nav>
          </Offcanvas.Body>
        </Navbar.Offcanvas>
      </Container>
    </Navbar>
  );
}
