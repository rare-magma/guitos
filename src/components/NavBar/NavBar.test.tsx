import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { vi } from "vitest";
import {
  budgetContextSpy,
  testBudget,
  testEmptyBudgetContext,
} from "../../setupTests";
import NavBar from "./NavBar";

describe("NavBar", () => {
  const onClone = vi.fn();
  const onExport = vi.fn();
  const onGoBack = vi.fn();
  const onGoHome = vi.fn();
  const onGoForward = vi.fn();
  const onImport = vi.fn();
  const onNew = vi.fn();
  const onRemove = vi.fn();
  const onRename = vi.fn();
  const onSelect = vi.fn();
  const comp = (
    <NavBar
      onClone={onClone}
      onExport={onExport}
      onGoBack={onGoBack}
      onGoHome={onGoHome}
      onGoForward={onGoForward}
      onImport={onImport}
      onNew={onNew}
      onRemove={onRemove}
      onRename={onRename}
      onSelect={onSelect}
    />
  );

  beforeEach(() => {
    render(comp);
  });

  it("matches snapshot", () => {
    expect(comp).toMatchSnapshot();
  });

  it("renders initial state", async () => {
    expect(screen.getByText("2023-03")).toBeInTheDocument();

    const newButton = screen.getAllByRole("button", { name: "new budget" });
    await userEvent.click(newButton[0]);
    await userEvent.click(newButton[0]);
    await userEvent.click(newButton[0]);

    expect(screen.getByPlaceholderText("Search...")).toBeInTheDocument();
    expect(screen.getByLabelText("go to older budget")).toBeInTheDocument();
    expect(screen.getByLabelText("go to newer budget")).toBeInTheDocument();
  });

  it("triggers onGo* when back/fwd buttons are pressed", async () => {
    await userEvent.click(screen.getByLabelText("go to older budget"));
    expect(onGoBack).toBeCalledTimes(1);
    onGoBack.mockClear();

    await userEvent.click(screen.getByLabelText("go to newer budget"));
    expect(onGoForward).toBeCalledTimes(1);
    onGoForward.mockClear();
  });

  it("triggers onGo* when back/fwd shortcuts are pressed", async () => {
    await userEvent.type(screen.getByTestId("header"), "{pagedown}");
    expect(onGoBack).toBeCalledTimes(1);

    await userEvent.type(screen.getByTestId("header"), "{home}");
    expect(onGoHome).toBeCalledTimes(1);

    await userEvent.type(screen.getByTestId("header"), "{pageup}");
    expect(onGoForward).toBeCalledTimes(1);
  });

  it("triggers onClone when clone button is pressed", async () => {
    await userEvent.click(screen.getByLabelText("clone budget"));
    expect(onClone).toBeCalledTimes(1);
  });

  it("triggers onImport when import button is pressed", async () => {
    await userEvent.click(screen.getByLabelText("import budget"));
    await userEvent.upload(
      screen.getByTestId("import-form-control"),
      testBudget as unknown as File,
    );
    expect(onImport).toBeCalledTimes(1);
  });

  it("triggers onExport when export json button is pressed", async () => {
    await userEvent.click(screen.getByLabelText("export budget"));
    await userEvent.click(
      screen.getByRole("button", { name: "export budget as json" }),
    );
    expect(onExport).toBeCalledWith("json");
  });

  it("triggers onExport when export csv button is pressed", async () => {
    await userEvent.click(screen.getByLabelText("export budget"));
    await userEvent.click(
      screen.getByRole("button", { name: "export budget as csv" }),
    );
    expect(onExport).toBeCalledWith("csv");
  });

  it("triggers onRename when user changes budget name input", async () => {
    await userEvent.type(screen.getByDisplayValue("2023-03"), "change name");

    expect(onRename).toBeCalledWith("2023-03change name");
    expect(screen.getByDisplayValue("2023-03change name")).toBeInTheDocument();
  });

  it("triggers onRemove when user clicks delete budget button", async () => {
    await userEvent.click(screen.getByLabelText("delete budget"));
    await userEvent.click(
      screen.getByRole("button", { name: "confirm budget deletion" }),
    );

    expect(onRemove).toBeCalledWith("035c2de4-00a4-403c-8f0e-f81339be9a4e");
  });

  it("opens instructions in new tab", async () => {
    const instructionsButton = screen.getByLabelText(
      "open instructions in new tab",
    );
    await userEvent.click(instructionsButton);
    expect(instructionsButton).toHaveAttribute(
      "href",
      "https://github.com/rare-magma/guitos#getting-started",
    );
  });

  it("opens guitos repo in new tab", async () => {
    budgetContextSpy.mockReturnValue(testEmptyBudgetContext);
    render(comp);
    const guitosButton = screen.getByLabelText("open guitos repository");
    await userEvent.click(guitosButton);
    expect(guitosButton).toHaveAttribute(
      "href",
      "https://github.com/rare-magma/guitos",
    );
  });
});
