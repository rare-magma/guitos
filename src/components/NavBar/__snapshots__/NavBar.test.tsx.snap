// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavBar > matches snapshot 1`] = `<NavBar />`;
